Sharing map variable-height trees

danpoe opened the pull request:
This changes the sharing map such that not all leafs are stored at the same depth (which was depth 6 previously), but that the tree grows dynamically as elements are inserted.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

danpoe committed:
Sharing map variable height trees

Previously the sharing map was a fixed-height tree (with height 6 by default).
This changes the sharing map to a variable-height tree.

danpoe committed:
Use key to search in add_item_if_not_shared()

When the method get_delta_view() encounters a leaf node in the first map, it
searches for a leaf with the same key in the corresponding subtree of the second
map. Previously all the nodes in the subtree were visited. This commit changes
the traversal such that the hash code of a key is used to determine the position
of the leaf in the second map.

danpoe committed:
Remove obsolete not_found_msg

danpoe committed:
Remove the tracking of tree node depth in the sharing map

When the sharing map was a fixed-height tree, the type of a node was determined
by its depth in the tree. Therefore, the node depth was tracked when traversing
the tree. Now that the sharing map is a variable-height tree, the methods
is_internal() and is_container() are used to determine the type of a node.

danpoe commented:
On a set of methods from Apache Tika, the sum of the runtimes of the sharing map methods (timing just those) improved by 22%. Overall symex was 4% faster.

smowton commented on src/util/sharing_map.h:
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
   iterate(n, f);
 }
 
+SHARING_MAPT(void)::add_item_if_not_shared(
```
Document this method

smowton commented on src/util/sharing_map.h:
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
+      if(ip->shares_with(container))
+        return;
+
+      const leaft &l1 = container.get_container().front();
```
Suggest making `l`, `ll`, `l1` more descriptive, they're pretty easy to muddle as it stands

smowton commented on src/util/sharing_map.h:
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
+      const leaf_listt &ll = ip->get_container();
+      SM_ASSERT(!ll.empty());
+
+      for(const auto &l : ll)
```
Comment explaining this list? It looks like you're sure `container` doesn't have multiple entries but `ip->get_container()` might do, and if it does and an entry is shared that means all further entries are too?

danpoe replied:
Yes, `add_item_if_not_shared()` is called by `get_delta_view()` only for containers which only have a single element. If we know that the leaf in the first container has the same key as a leaf in the second container (either because they share or the keys compare equal), then since the maps cannot contain duplicate keys, we can ignore all further keys in the second container since they will be different from the key of the leaf in the first container. We can ignore the further leafs in the second container since `map1.get_delta_view(map2, ...)` does not include items that are only in `map2`.

smowton commented on src/util/sharing_map.h:
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
+        if(l1.shares_with(l))
+          return;
+
+        if(l1.get_key() == l.get_key())
```
Why don't we do anything when differing keys occur in `container` and `ip->get_container()`? Is that impossible for some reason?

danpoe replied:
Ah yes there's a case missing for when we don't find a matching key. The bug would only appear when there are hash collisions. I'll fix it and add a test.

smowton commented on src/util/sharing_map.h:
```diff
@@ -917,21 +995,28 @@ ::get_delta_view(
-
-  return ip;
+  UNREACHABLE;
+  return nullptr;
```
I think some compilers will complain about this unreachable statement

smowton commented on src/util/sharing_map.h:
```diff
@@ -1001,13 +1098,138 @@ SHARING_MAPT(void)::erase(const key_type &k)
   num--;
 }
 
+SHARING_MAPT2(, innert *)::migrate(
```
Document this methid

smowton commented on unit/util/sharing_map.cpp:
```diff
@@ -126,6 +131,163 @@ void sharing_map_internals_test()
     REQUIRE(marked.size() >= 2);
   }
+
+  SECTION("single map shape")
```
Describe this in some more detail, what is shape?

smowton commented on unit/util/sharing_map.cpp:
```diff
@@ -214,15 +214,14 @@ void sharing_map_internals_test()
       REQUIRE(delta_view.size() == 2);
 
-      REQUIRE(!delta_view[0].in_both);
+      REQUIRE(!delta_view[0].is_in_both_maps());
```
Squash this commit

smowton commented on src/util/sharing_map.h:
```diff
@@ -1283,7 +1288,7 @@ SHARING_MAPT2(optionalt<std::reference_wrapper<const, mapped_type>>)::find(
 SHARING_MAPT(const std::string)::not_found_msg="key not found";
 
-SHARING_MAPT(const std::size_t)::bits = 18;
+SHARING_MAPT(const std::size_t)::bits = 30;
```
Surely belongs in its own commit

danpoe replied:
Done

tautschnig commented on src/util/sharing_map.h:
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
+      for(const auto &item : m)
+      {
+        const innert *i = &item.second;
+        stack.push(i);
```
I don't think `i` adds any value here, I'd just do `stack.push(&item.second);`

tautschnig commented on src/util/sharing_map.h:
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
+      for(const auto &l : ll)
+      {
+        if(l1.shares_with(l))
+          return;
```
I would have expected `continue` - why is it ok to return?

danpoe replied:
That's because if `l1` shares with `l`, then since the maps don't contain duplicate keys, all further keys in the container must be different from `l1.get_key()`. Since `map1.get_delta_view(map2, ...)` does not include items that are only in `map2`, we can return immediately.

tautschnig commented on src/util/sharing_map.h:
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
+          // element is in both maps and not shared
+          delta_view.push_back(
+            {true, l1.get_key(), l1.get_value(), l.get_value()});
+          return;
```
Again, I would have expected `continue` instead of returning.

danpoe replied:
Similar to above, since we have found a matching key, all other keys will be different and don't have to be included in the `delta_view`.

tautschnig commented on src/util/sharing_map.h:
```diff
@@ -1001,13 +1098,138 @@ SHARING_MAPT(void)::erase(const key_type &k)
+  innert *ip)
+{
+  SM_ASSERT(step < steps - 1);
+  SM_ASSERT(ip != nullptr);
```
If it must not be a null pointer, why not pass in a reference?

tautschnig commented on src/util/sharing_map.h:
```diff
@@ -1001,13 +1098,138 @@ SHARING_MAPT(void)::erase(const key_type &k)
+
+  // Add internal node
+  ip = ip->add_child(bit_last);
+  SM_ASSERT(ip->empty());
```
I would just declare a local pointer variable instead of re-using the `ip` parameter.

danpoe replied:
Changed

tautschnig commented on src/util/sharing_map.h:
```diff
@@ -1001,13 +1098,138 @@ SHARING_MAPT(void)::erase(const key_type &k)
+
+  SM_ASSERT(i < steps);
+
+  do
```
How about using a `for` loop instead: `for(std::size_t i = step + 1; i < steps; ++i)`?

danpoe replied:
For the first iteration we don't need to check the condition so a do-while is probably fine.

tautschnig commented on src/util/sharing_map.h:
```diff
@@ -1001,13 +1098,138 @@ SHARING_MAPT(void)::erase(const key_type &k)
+
+  leaft leaf_copy(as_const(&container_copy)->get_container().front());
+  leaf_listt &c = ip->get_container();
+  c.push_front(leaf_copy);
```
`c` doesn't seem to add much value: `ip->get_container().push_front(leaf_copy);`

tautschnig commented on src/util/sharing_map.h:
```diff
@@ -922,6 +923,7 @@ ::get_delta_view(
         if(!child.shares_with(*ip2))
         {
           stack.push(stack_itemt(&child, ip2));
+          level_stack.push(0xff);
```
Maybe use `optionalt`, or else name that constant.

danpoe committed:
Increase maximum sharing map tree height from 6 to 10

Now that the sharing map tree grows dynamically as elements are inserted, a
higher value for the maximum tree height can be used. The maximum tree height is
determined by bits / chunk, with bits being the number of bits of the hash code
used, and chunk being the number of bits used to determine the child node of an
inner node. For example, if chunk = 3, then a node can have at most 8 children.
Then if bits = 30, bits / chunk = 10 and the maximum tree height is 10. The
constant bits should be divisible by chunk.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 38c8388639e406c7e79109f6731584872a870fec).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/111936752
